users/jonavin/jonavin.c

```diff
@@ -0,0 +1,248 @@
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include QMK_KEYBOARD_H
+
+// DEFINE MACROS
+#define ARRAYSIZE(arr)  sizeof(arr)/sizeof(arr[0])
+
+
+// LAYERS
+__attribute__ ((weak))
+enum custom_layers {
+    _BASE,
+    _FN1,
+    _LOWER,
+    _RAISE,
+};
+
+// KEYCODES
+__attribute__ ((weak))
+enum custom_keycodes {
+  KC_00 = SAFE_RANGE,
+  ENCFUNC,
+  KC_WINLCK,    //Toggles Win key on and off
+  RGB_TOI,   // Timeout idle time up
+  RGB_TOD,   // Timeout idle time down
+};
+
+#define KC_CAD	LALT(LCTL(KC_DEL))
+#define KC_AF4	LALT(KC_F4)
+#define KC_TASK	LCTL(LSFT(KC_ESC))
+
+// Win Key Lock
+__attribute__ ((weak)) bool _isWinKeyDisabled = false;
+
+#ifdef LSFT_CAPSLOCK_ENABLE
+    // Tap Dance Definitions
+    enum custom_tapdance {
+        TD_LSFT_CAPSLOCK,
+    };
+
+    qk_tap_dance_action_t tap_dance_actions[] = {
+    // Tap once for shift, twice for Caps Lock
+        [TD_LSFT_CAPSLOCK] = ACTION_TAP_DANCE_DOUBLE(KC_LSFT, KC_CAPS),
+    };
+
+    #define KC_LSFTCAPS TD(TD_LSFT_CAPSLOCK)
+#else // regular Shift
+    #define KC_LSFTCAPS KC_LSFT
+#endif // LSFT_CAPSLOCK_ENABLE
+
+
+
+// TIMEOUTS
+#ifdef IDLE_TIMEOUT_ENABLE
+    #define TIMEOUT_THRESHOLD_DEFAULT   5    // default timeout minutes
+    #define TIMEOUT_THRESHOLD_MAX       140  // upper limits (2 hours and 10 minutes -- no rgb indicators above this value)
+    static uint16_t timeout_timer = 0;
+    static uint16_t timeout_counter = 0;  //in minute intervals
+    static uint16_t timeout_threshold = TIMEOUT_THRESHOLD_DEFAULT;
+
+    void timeout_reset_timer(void) {
+        timeout_timer = timer_read();
+        timeout_counter = 0;
+    };
+
+    void timeout_update_threshold(bool increase) {
+        if (increase && timeout_threshold < TIMEOUT_THRESHOLD_MAX) timeout_threshold++;
+        if (!increase && timeout_threshold > 0) timeout_threshold--;
+    };
+
+    void timeout_tick_timer(void) {
+        if (timeout_threshold > 0) {
+            if (timer_elapsed(timeout_timer) >= 60000) { // 1 minute tick
+                timeout_counter++;
+                timeout_timer = timer_read();
+            }
+            #ifdef RGB_MATRIX_ENABLE
+                if (timeout_threshold > 0 && timeout_counter >= timeout_threshold) {
+                    rgb_matrix_disable_noeeprom();
+                }
+            #endif
+        } // timeout_threshold = 0 will disable timeout
+    }
+
+    __attribute__ ((weak))  void matrix_scan_user(void) {
+        timeout_tick_timer();
+    };
```

Comment:
Something like this may be better: 
```suggestion
__attribute__((weak)) void matrix_scan_keymap(void) {}

void matrix_scan_user(void) {
    timeout_tick_timer();
    matrix_scan_keymap();
}
```